--- a/utils/kamctl/kamdbctl.base
+++ b/utils/kamctl/kamdbctl.base
@@ -33,20 +33,20 @@
 
 # Used by dbtext and db_berkeley to define tables to be created, used by
 # postgres to do the grants
-STANDARD_TABLES=${STANDARD_TABLES:-version acc dbaliases domain domain_attrs
+STANDARD_TABLES=${STANDARD_TABLES:-'version acc dbaliases domain domain_attrs
 		grp uri speed_dial lcr_gw lcr_rule lcr_rule_target pdt subscriber
 		location location_attrs re_grp trusted address missed_calls
 		usr_preferences aliases silo dialog dialog_vars dispatcher dialplan
-		acc_cdrs topos_d topos_t}
-EXTRA_TABLES=${EXTRA_TABLES:-imc_members imc_rooms cpl sip_trace domainpolicy
+		acc_cdrs topos_d topos_t'}
+EXTRA_TABLES=${EXTRA_TABLES:-'imc_members imc_rooms cpl sip_trace domainpolicy
 		carrierroute carrier_name domain_name carrierfailureroute userblacklist
 		globalblacklist htable purplemap uacreg pl_pipes mtree mtrees
 		sca_subscriptions mohqcalls mohqueues rtpproxy rtpengine
-		dr_gateways dr_rules dr_gw_lists dr_groups}
-PRESENCE_TABLES=${PRESENCE_TABLES:-presentity active_watchers watchers xcap
-		pua rls_presentity rls_watchers}
-DBUID_TABLES=${UID_TABLES:-uid_credentials uid_domain uid_domain_attrs
-		uid_global_attrs uid_uri uid_uri_attrs uid_user_attrs}
+		dr_gateways dr_rules dr_gw_lists dr_groups'}
+PRESENCE_TABLES=${PRESENCE_TABLES:-'presentity active_watchers watchers xcap
+		pua rls_presentity rls_watchers'}
+DBUID_TABLES=${UID_TABLES:-'uid_credentials uid_domain uid_domain_attrs
+		uid_global_attrs uid_uri uid_uri_attrs uid_user_attrs'}
 
 # SQL definitions
 # If you change this definitions here, then you must change them
@@ -69,18 +69,18 @@
 SED=${SED:-sed}
 
 # define what modules should be installed
-STANDARD_MODULES=${STANDARD_MODULES:-standard acc lcr domain group
+STANDARD_MODULES=${STANDARD_MODULES:-'standard acc lcr domain group
 		permissions registrar usrloc msilo alias_db uri_db speeddial
-		avpops auth_db pdt dialog dispatcher dialplan topos}
+		avpops auth_db pdt dialog dispatcher dialplan topos'}
 
-PRESENCE_MODULES=${PRESENCE_MODULES:-presence rls}
+PRESENCE_MODULES=${PRESENCE_MODULES:-'presence rls'}
 
-EXTRA_MODULES=${EXTRA_MODULES:-imc cpl siptrace domainpolicy carrierroute
+EXTRA_MODULES=${EXTRA_MODULES:-'imc cpl siptrace domainpolicy carrierroute
 		drouting userblacklist htable purple uac pipelimit mtree sca mohqueue
-		rtpproxy rtpengine}
+		rtpproxy rtpengine'}
 
-DBUID_MODULES=${UID_MODULES:-uid_auth_db uid_avp_db uid_domain uid_gflags
-		uid_uri_db}
+DBUID_MODULES=${UID_MODULES:-'uid_auth_db uid_avp_db uid_domain uid_gflags
+		uid_uri_db'}
 
 ############################################################
 
