diff --git a/src/modules/ims_dialog/dlg_hash.c b/src/modules/ims_dialog/dlg_hash.c
index 999ae01..4fe4aae 100644
--- a/src/modules/ims_dialog/dlg_hash.c
+++ b/src/modules/ims_dialog/dlg_hash.c
@@ -17,7 +17,8 @@
 #include "dlg_profile.h"
 #include "dlg_handlers.h"
 #include "dlg_db_handler.h"
-#include <execinfo.h>
+
+// #include <execinfo.h>
 
 #define MAX_LDG_LOCKS  2048
 #define MIN_LDG_LOCKS  2
@@ -70,6 +71,7 @@ static int dlg_hash_size_out = 4096;
 		}\
 	}while(0)
 
+#ifdef ALPINE_SUPPORTS_BACKTRACE
 inline static int backtrace2str(char* buf, int size)
 {
         void* bt[32];
@@ -87,6 +89,7 @@ inline static int backtrace2str(char* buf, int size)
         }
         return 0;
 }
+#endif
 
 /*!
  * \brief Initialize the global dialog table
